### PR TITLE
add support for profile with dbapi

### LINKED ISSUE #743: add support for `%sqlcmd profile` with DBAPI drivers
`%sqlcmd profile` should work with DBAPIConnections: https://github.com/ploomber/jupysql/blob/7c1b5eaed0a7a7b787adecd2c91388146efda6e3/src/sql/connection/connection.py#L564

image

### PR DESCRIPTION
**Pull request by bbeat2782**
## Describe your changes

Modify code so that `%sqlcmd profile` also works with DBAPIConnections

## Issue number

Closes #743 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--785.org.readthedocs.build/en/785/

### CODE REVIEW

File: src/sql/inspect.py [hunk 1]
```diff
@@ -422,9 +422,14 @@ def __init__(self, table_name, schema=None) -> None:
             warning_background = "white"
             warning_title = ""
 
-        database = ConnectionManager.current.url
-        db_driver = ConnectionManager.current._get_database_information()["driver"]
-        if "duckdb" in database:
+        current = ConnectionManager.current
```
Comment (edublancas) [inline src/sql/inspect.py:425]: I'm confused about this change, why is it needed?

also, you can access current.dialect, no need to add the `if` statement, check the developer docs for more details

File: src/sql/magic_cmd.py [hunk 2]
```diff
@@ -65,6 +64,8 @@ def _validate_execute_inputs(self, line):
             command, others = split[0].strip(), split[1:]
 
             if command in AVAILABLE_SQLCMD_COMMANDS:
+                if command not in ["profile"]:
+                    util.support_only_sql_alchemy_connection("%sqlcmd")
```
Comment (edublancas) [inline src/sql/magic_cmd.py:68]: looks like this will raise an error whenever we execute something that's not `profile`? why?

File: src/tests/test_magic_cmd.py [hunk 3]
```diff
@@ -170,10 +185,18 @@ def test_columns_with_schema(ip, tmp_empty):
...
 
 
-def test_table_profile(ip, tmp_empty):
-    ip.run_cell(
+@pytest.mark.parametrize(
+    "conn",
+    [
+        ("%sql sqlite://"),
+        ("%sql sqlite_conn"),
+    ],
```
Comment (edublancas) [inline src/tests/test_magic_cmd.py:193]: you can add all the connections to your fixture with an alias and then pass `%sql connectionalias` to switch to it, this is clear than that approach you're taking. also provide more clear variable names, because `sqlite_conn` doesn't immediately tell me that this is a DBAPI connection

File: src/tests/test_magic_cmd.py [hunk 4]
```diff
@@ -319,16 +350,67 @@ def test_table_schema_profile(ip, tmp_empty):
...
 
-def test_table_profile_warnings_styles(ip, tmp_empty):
-    ip.run_cell(
+def test_table_schema_profile_dbapi(ip_with_dbapi, tmp_empty):
+    sqlite_conn_b = sqlite3.connect("b.db")
+    ip_with_dbapi.push({"sqlite_conn_b": sqlite_conn_b})
+
+    ip_with_dbapi.run_cell(
+        """%%sql sqlite_conn_b
+CREATE TABLE t (n FLOAT);
```
Comment (edublancas) [inline src/tests/test_magic_cmd.py:359]: provide more descriptive names to table names

File: src/tests/test_magic_cmd.py [hunk 5]
```diff
@@ -319,16 +350,67 @@ def test_table_schema_profile(ip, tmp_empty):
...
+        "mean": ["22.0000"],
+        "min": ["11.0"],
+        "max": ["33.0"],
+        "std": ["11.0000"],
+        "unique": ["3"],
+        "freq": [math.nan],
+        "top": [math.nan],
+    }
+
+    out = ip_with_dbapi.run_cell("%sqlcmd profile -t t --schema b_schema").result
```
Comment (edublancas) [inline src/tests/test_magic_cmd.py:384]: when writing tests, it's better to provide the full name of each argument (`--table` instead of `-t`)

**Review by edublancas — CHANGES_REQUESTED** [review]

File: src/tests/test_magic_cmd.py (hunk 5 above)
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR387)

File: src/tests/test_magic_cmd.py (hunk 4 above)
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR361)

File: src/tests/test_magic_cmd.py [hunk 6]
```diff
@@ -319,16 +350,67 @@ def test_table_schema_profile(ip, tmp_empty):
             assert cell == str(expected[profile_metric][0])
 
 
-def test_table_profile_warnings_styles(ip, tmp_empty):
-    ip.run_cell(
+def test_table_schema_profile_dbapi(ip_with_dbapi, tmp_empty):
```
[the remark this answers is not in the document]
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR355)

Changed to `test_sqlcmd_profile_with_schema_argument_and_dbapi`

File: src/tests/test_magic_cmd.py (hunk 3 above)
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR68-R80): Changing the fixture. Now aliases are `duckdb_sqlalchemy`, `sqlite_sqlalchemy`, `duckdb_dbapi`, and `sqlite_dbapi`.

Also [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR193-R194) parametrize arguments.

File: src/tests/test_magic_cmd.py [hunk 7]
```diff
@@ -219,10 +242,18 @@ def test_table_profile(ip, tmp_empty):
     assert "position: sticky;" in out._table_html
 
 
-def test_table_profile_with_stdev(ip, tmp_empty):
-    ip.run_cell(
+@pytest.mark.parametrize(
+    "conn",
+    [
+        ("%sql duckdb://"),
+        ("%sql duckdb_conn"),
```
[the remark this answers is not in the document]
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR250-R251)

Also fixed [here](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR250-R251), [here](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR438-R439), and [here](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR467-R468) that have the same issue

File: src/sql/magic_cmd.py (hunk 2 above)
Comment (bbeat2782) [reply src/sql/magic_cmd.py:68]: `util.support_only_sql_alchemy_connection` raises error when the current connection is not SQLAlchemy ([code](https://github.com/ploomber/jupysql/blob/master/src/sql/util.py#L255)). This is included originally since dbapi connection does not support `%sqlcmd` arguments.

Since this issue is about adding support for profile, I moved `util.support_only_sql_alchemy_connection` inside the if statement so that it doesn't raise error when a user runs `%sqlcmd profile` but still raises the error when other commands are used.

File: src/sql/inspect.py (hunk 1 above)
Comment (bbeat2782) [reply src/sql/inspect.py:425]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-644b34a1c0db055c7b331734967239a598bd8f6e07cf50912c919bf70d243f3dR426)

Removed the if statement and changed to use `current.dialect` instead of `current.url` and `current._dialect`.

The change is there because: We need to distinguish whether a given connection is duckdb or not so that we can display an appropriate message that shows which statistics are not available. However, with `current.url` dbapi connection always returns None.

File: src/tests/test_magic_cmd.py [hunk 8]
```diff
@@ -63,6 +65,21 @@ def ip_snippets(ip):
     yield ip
 
 
+@pytest.fixture
+def ip_with_dbapi(ip):
```
Comment (edublancas) [inline src/tests/test_magic_cmd.py:69]: clearer name since this also contains many types of connections, not only dbapi

```suggestion
def ip_with_connections(ip_empty):
```

also, switched the fixture to `ip_empty`, since `ip` contains sample data that you don't need for this test

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,6 +2,7 @@
 
 ## 0.10.0dev
 
+* [Feature] Add DBAPI support for `%sqlcmd profile`(#743)
```
Comment (edublancas) [inline CHANGELOG.md:5]:
```suggestion
* [Feature] Add `%sqlcmd profile` support for DBAPI connections (#743)
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -19,7 +20,8 @@
 * [Fix] Fix boxplot for duckdb native ([#728](https://github.com/ploomber/jupysql/issues/728))
 * [Fix] Fix error when using SQL Server with pyodbc that caused queries to fail due to multiple open result sets
 * [Fix] Improves performance when converting DuckDB results to `pandas.DataFrame`
-* [Feature] Add DBAPI support for `%sqlcmd profile`(#743)
+* [Fix] Fixes a bug when converting a CTE stored with `--save` into a `pandas.DataFrame` via `.DataFrame()`
+* [Doc] Add Redshift tutorial
```
Comment (edublancas) [inline CHANGELOG.md:24]: looks like a merge went bad? this entries look unrelated to your PR

**Review by edublancas — CHANGES_REQUESTED** [review]
minor comments

File: CHANGELOG.md (hunk 10 above)
Comment (bbeat2782) [reply CHANGELOG.md:24]: Where did you find this change? It doesn't appear on my end

File: CHANGELOG.md (hunk 9 above)
Comment (bbeat2782) [reply CHANGELOG.md]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR5)

File: src/tests/test_magic_cmd.py (hunk 8 above)
Comment (bbeat2782) [reply src/tests/test_magic_cmd.py]: [fixed](https://github.com/ploomber/jupysql/pull/785/files#diff-9188e75e824219dd191cdc62d6d19f2b15c4bd0cf564abe3646ea907c32cd61dR69-R80)

File: CHANGELOG.md (hunk 10 above)
Comment (edublancas) [reply CHANGELOG.md]: weird. I don't see it anymore. marking as fixed

**Review by edublancas — APPROVED** [review]
